From d7755624eefdbe3691788659c96a93a8481035ab Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Thu, 16 May 2024 04:01:27 +0000
Subject: [PATCH] pfe: phylink: set phy interface mode to SGMII when in 2.5G

On RDB3 board, the default emac1 is connected with AQR113C phy, and the
supported speed is up to 2.5G(SGMII enaled). But when it runs in 2.5G,
if we bring the PFE down and then bring it back up, we will encounter the
following error:

  # ifconfig pfe1 down
  pfeng 46000000.pfe pfe1: Link is Down
  # ifconfig pfe1 up
  pfeng 46000000.pfe pfe1: validation of 2500base-x with support 00000000,00018000,000e706c and advertisement 00000000,00008000,0000606c failed: -EINVAL
  pfeng 46000000.pfe pfe1: ERR: (DRIVER) event 1 - Driver runtime error: [pfeng-phylink.c:386] could not attach PHY: -22
  pfeng 46000000.pfe pfe1: ERR: (DRIVER) event 1 - Driver runtime error: [pfeng-netif.c:147] Error connecting to the phy: -22
  SIOCSIFFLAGS: Invalid argument

The reason is that, when pfe1 is running in 2.5G, the related phy device
will set the interface mode to PHY_INTERFACE_MODE_2500BASEX, which is not
valid in pfeng_phylink_validate(), and if we try to add the above
PHY_INTERFACE_MODE_2500BASEX to the valid list, like SGMII and R/G/MII modes,
the default speed of pfe1 will be 1G instead, which is not reasonable.

So let's set it to PHY_INTERFACE_MODE_SGMII just in one own code block,
to fix the issue and reduce the other side effects.

Upstream-Status: Pending

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 sw/linux-pfeng/pfeng-phylink.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/sw/linux-pfeng/pfeng-phylink.c b/sw/linux-pfeng/pfeng-phylink.c
index 4092c66..df90e9a 100644
--- a/sw/linux-pfeng/pfeng-phylink.c
+++ b/sw/linux-pfeng/pfeng-phylink.c
@@ -105,6 +105,17 @@ static void pfeng_phylink_validate(struct phylink_config *config, unsigned long
 	int max_speed = priv->emac[netif->cfg->phyif_id].max_speed;
 	int an_serdes_speed = priv->emac[netif->cfg->phyif_id].serdes_an_speed;
 
+
+	/* the default emac1 connected with AQR113C phy, and the supported
+	 * speed is up to 2.5G(SGMII enaled). But when it runs in 2.5G, the
+	 * state->interface will be set to PHY_INTERFACE_MODE_2500BASEX by the
+	 * phydevice, but as following codes show, only SGMII and R/G/MII modes
+	 * are valid, so let's set it to PHY_INTERFACE_MODE_SGMII instead.
+	 */
+	if (state->interface == PHY_INTERFACE_MODE_2500BASEX) {
+		state->interface = PHY_INTERFACE_MODE_SGMII;
+	}
+
 	/* We only support SGMII and R/G/MII modes */
 	if (state->interface != PHY_INTERFACE_MODE_NA &&
 		state->interface != PHY_INTERFACE_MODE_SGMII &&
-- 
2.34.1

